src/xmipp/applications/programs/mpi_nma_alignment_vol/mpi_nma_alignment_vol_main.cpp

```diff
@@ -0,0 +1,140 @@
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+ * 02111-1307  USA
+ *
+ *  All comments concerning this program package may be sent to the
+ *  e-mail address 'xmipp@cnb.uam.es'
+ ***************************************************************************/
+//#include <mpi.h>
+#include <parallel/xmipp_mpi.h>
+#include <reconstruction/nma_alignment_vol.h>
+
+
+/** Class to perfom the NMA Alignment with  MPI parallelization */
+class MpiProgNMAVol: public ProgNmaAlignmentVol
+{
+private:
+    MpiNode *node;
+    MpiTaskDistributor *distributor;
+    std::vector<size_t> imgsId;
+    MpiFileMutex *fileMutex;
+
+public:
+    /** Empty Constructor */
+    MpiProgNMAVol()
+    {
+    	node=NULL;
+    	distributor=NULL;
+    	fileMutex=NULL;
+    }
+
+    /** Destructor */
+    ~MpiProgNMAVol()
+    {
+        delete node;
+        delete fileMutex;
+        delete distributor;
+    }
+
+    /** Redefine read to initialize MPI environment */
+    void read(int argc, char **argv)
+    {
+        node = new MpiNode(argc, argv);
+        if (!node->isMaster())
+        	verbose=0;
+        fileMutex = new MpiFileMutex(node);
+        ProgNmaAlignmentVol::read(argc, argv);
+    }
+    /** main body */
+    void createWorkFiles()
+    {
+        //Master node should prepare some stuff before start working
+        MetaData &mdIn = *getInputMd(); //get a reference to input metadata
+
+        if (node->isMaster())
+        {
+            ProgNmaAlignmentVol::createWorkFiles();
+            mdIn.write(fnOutDir + "/nmaTodo.xmd");
+        }
+        node->barrierWait();//Sync all before start working
+        mdIn.read(fnOutDir + "/nmaTodo.xmd");
+        mdIn.findObjects(imgsId);//get objects ids
+        rangen = node->rank;
+        distributor = new MpiTaskDistributor(mdIn.size(), 1, node);
+    }
+    //Only master do starting progress bar stuff
+    void startProcessing()
+    {
+        if (node->isMaster())
+            ProgNmaAlignmentVol::startProcessing();
+    }
+    //Only master show progress
+    void showProgress()
+    {
+        if (node->isMaster())
+            ProgNmaAlignmentVol::showProgress();
+    }
+    //Now use the distributor to grasp images
+    bool getImageToProcess(size_t &objId, size_t &objIndex)
+    {
+        size_t first, last;
+        bool moreTasks = distributor->getTasks(first, last);
+
+        if (moreTasks)
+        {
+            time_bar_done = first + 1;
```

Comment:
I sense a race condition here. Can you check it?

Reply:
You may be true, I don't see the progress of the program when mpi is used. This was never reported as an issue because the time_bar doesn't effect the functionality of the program.
Do you have suggestions on how to solve it if exists? Thanks.

Reply:
Have a look [here](https://www.modernescpp.com/index.php/prefer-locks-to-mutexes) and e.g. [here](https://en.cppreference.com/w/cpp/thread/unique_lock)

Reply:
But don't worry about this one, we should deal with it properly at the level of all Xmipp programs